src/universal/components/SignInPage/SignInPage.js

```diff
@@ -0,0 +1,222 @@
+ * also functions as the callback handler for the Auth0 OIDC response.
+ *
+ * @flow
+ */
+import type {Dispatch} from 'redux';
+import type {RouterHistory, Location} from 'react-router-dom';
+
+import React, {Component} from 'react';
+import Loadable from 'react-loadable';
+import {connect} from 'react-redux';
+import {Link, withRouter} from 'react-router-dom';
+
+import signinAndUpdateToken from 'universal/components/Auth0ShowLock/signinAndUpdateToken';
+import LoadableLoading from 'universal/components/LoadableLoading';
+import LoadingView from 'universal/components/LoadingView/LoadingView';
+import withAtmosphere from 'universal/decorators/withAtmosphere/withAtmosphere';
+
+import Header from './Header';
+import SignIn from './SignIn';
+import chunkArray from 'universal/utils/chunkArray';
+
+type Credentials = {
+  email: string,
+  password: string
+};
+
+type ParsedAuthResponse = {
+  idToken: string,
+  idTokenPayload: Object
+};
+
+type Props = {
+  atmosphere: Object,
+  dispatch: Dispatch<*>,
+  hasSession: boolean,
+  history: RouterHistory,
+  location: Location,
+  webAuth: Object
+};
+
+type State = {
+  loggingIn: boolean,
+  error: ?Error
+};
+
+const containerStyles = {
+  display: 'flex',
+  flexDirection: 'column'
+};
+
+const loadingErrorStyles = ({
+  display: 'flex',
+  flexDirection: 'column',
+  alignItems: 'center',
+  height: '100%'
+});
+
+// A naive implementation taking a querystring-formatted string (beginning in '?'),
+// and returning an object of the key/value pairs.  Will not work for null-valued
+// keys, but works for our own `?redirectTo=/path/to/page` contract.
+const parseSearch = (search: string): Object => {
+  if (!search) { return {}; }
+  const assocArray = chunkArray(search.slice(1).split(/=|&/), 2);
+  return assocArray.reduce((acc, [key, val]) => ({...acc, [key]: val}), {});
+};
+
+// Note that auth0-js seems to break SSR.  Therefore we dynamically import it
+// over the network.
+const LoadableSignInPage = Loadable({
+  loader: () => import('auth0-js'),
+  loading: LoadableLoading,
+  render: (auth0, props) => {
+    const webAuth = new auth0.WebAuth({
+      domain: __AUTH0_DOMAIN__,
+      clientID: __AUTH0_CLIENT_ID__,
+      redirectUri: window.location.href,
+      scope: 'openid rol tms bet'
+    });
+    return <SignInPage webAuth={webAuth} {...props} />;
+  }
+});
+
+class SignInPage extends Component<Props, State> {
+  state = {
+    error: null,
+    loggingIn: false
+  };
+
+  componentWillMount() {
+    this.maybeRedirectToApp();
+    this.maybeCaptureAuthResponse();
+  }
+
+  componentDidUpdate() {
+    this.maybeRedirectToApp();
+  }
+
+  getHandlerForThirdPartyAuth = (auth0Connection: string) => () => {
+    this.props.webAuth.authorize({
+      connection: auth0Connection,
+      responseType: 'token'
+    });
+  };
+
+  maybeRedirectToApp = () => {
+    if (this.props.hasSession) {
+      const parsedSearch = parseSearch(this.props.location.search);
```

Comment:
we previously stored this in redux to keep the url clean. if you like this better i'm cool with it, but we should remove `nextUrl` from the authDuck then

Reply:
Oh, well if this behavior is already implemented, I can simply remove it.

Reply:
Actually, yeah, I prefer the stateless querystring param approach. Removing the nextUrl stuff.